File: day1/TS/src/day1.ts
import fs from "fs";

export function part1(filepath: string): number{
    const data = fs.readFileSync(filepath,"utf-8");
    const lines = data.trim().split('\n')
    const left: number[] = []
    const right: number[] = []
    for (const line of lines){
        /*
         * Splittar och stannar när det inte finns whitespace mer så när den når nästa siffra.
         * Sen konverterar jag string elementen till tal med map(Number) om det skulle varit något som inte är ett nummer i filen hade det kommit tillbaka som NaN eller på raden i detta fallet.
        */
        const [leftNum,rightNum] = line.split(/\s+/).map(Number);
        left.push(leftNum)
        right.push(rightNum)
    }
    /*
    * Sort hade jämför talen som strängar och inte sorterat dom korrekt.
    * Därför så specifiserar jag hur dom ska sorteras.
    * Så om listan är [3,1,4] så jämför den först 3 och 1 sedan 1 och 4
    * och eftersom att i det fösta exmpelet så blir det 3 - 1 som blir 2 och är positivt så flyttas 1 förre 3 osv (i detta fallet.) 
    */
    const leftSorted = [...left].sort((a,b) => a - b);
    const rightSorted = [...right].sort((a,b) => a - b);

    let totalDistance = 0;
    for (let i = 0; i < leftSorted.length; i++){
        totalDistance += Math.abs(leftSorted[i] - rightSorted[i])
    }
    return totalDistance
}

export function part2(filepath: string): number{
    const data = fs.readFileSync(filepath,"utf-8");
    const lines = data.trim().split('\n');
    const left: number[] = []
    const right: number[] =  []

    for(const line of lines){
        const [leftNumb,rightNumb] = line.split(/\s+/).map(Number);
        left.push(leftNumb);
        right.push(rightNumb);
    }
    const rightCounter = countOccurrences(right)
    let simScore = 0;
    for(const num of left){
        //Hämtar det ifall det finns och inte så ger vi det värdet 0.
        const occurrances = rightCounter[num] || 0;
        simScore += num*occurrances;
    }
    return simScore;
}
/*
* Funktion för att lägga ihopa talet och frekvensen av talet iform av en hashmap. Detta är vad Counter gör åt oss i python delen men det finns inte med i TS.
* Som exempel:
* const right = [4,3,5,3,9,3]
* const rightCounter = countOccurances(right);
* console.log(rightCounter) Ger outputen: {4:1,3:3,5:1,9:1}
*/
function countOccurrences(array: number[]): Record<number,number>{
    const counts: Record<number,number> = {};
    for(const number of array){
        counts[number] = (counts[number] || 0)+1;
    }
    return counts;
}

File: day1/rust/src/day1.rs
use std::fs;
use std::collections::HashMap;

pub fn part1(filepath: &str) -> i32 {
    //Läser in filen.
    let data = fs::read_to_string(filepath).expect("Kunde inte läsa filen");

    //Skapar vektorer för left och right. Använder Vector över Arc eftersom att detta ska inte delas över olika threads eller ägande.
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    for line in data.lines() {
        let numbers: Vec<i32> = line
        .split_whitespace()
        //För varje num så konverterar vi det till ett heltal.
        .map(|num| num.parse::<i32>()
        .expect("Kunde inte parsa"))
        .collect();
        
        left.push(numbers[0]);
        right.push(numbers[1]);
    }
    left.sort_unstable();
    right.sort_unstable();

    /*
    * Itererar över den vänsta listan och slår sedan ihopa den med en iteration av högerlistan.
    * Mappar sedan över båda listorna l och r som jag kallar dom och räknar ut den absoluta skillnaden.
    * Summerar och returnerar det.
    */
    let total_distance: i32 = left
    .iter()
    .zip(right.iter())
    .map(|(l,r)|(l-r).abs())
    .sum();

    total_distance
}

pub fn part2(filepath: &str) -> i32 {
    let data = fs::read_to_string(filepath).expect("Kunde inte läsa filen!");
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    for line in data.lines() {
        let numbers: Vec<i32> = line
        .split_whitespace()
        .map(|num| num.parse::<i32>().expect("Kunde inte parsa!"))
        .collect();
        
        left.push(numbers[0]);
        right.push(numbers[1]);
    }
    left.sort_unstable();
    right.sort_unstable();

    let mut right_counter: HashMap<i32,i32> = HashMap::new();
    
    for num in right {
        //Kollar ifall nyckel redan finns (nummert) ifall den inte gör det skapar den en nyckel med standard värdet 0 som sedan blir +1.
        *right_counter.entry(num).or_insert(0) += 1;
    }
    let mut sim_score = 0;
    for num in left {
        if let Some(count) = right_counter.get(&num){
            sim_score += num * count;
        }
    }

    sim_score

}